src/adaXT/decision_tree/decision_tree.pyx

```diff
@@ -176,25 +177,60 @@ class DecisionTree(BaseModel):
         if (not leaf_nodes):  # make sure that there are calculated observations
             raise ValueError("The tree has no leaf nodes")
 
-        n_rows = self.n_rows
-        matrix = np.zeros((n_rows, n_rows))
+        ht = {}
         for node in leaf_nodes:
-            indices = node.indices
-            for i in indices:
-                for j in indices:
-                    if scale:
-                        n_node = len(node.indices)
-                        matrix[i, j] += 1/n_node
-                    else:
-                        matrix[i, j] += 1
+            ht[node.id] = node.indices
+        return ht
+
+    def __tree_based_weights(self, hash1, hash2, size_X0, size_X1, scale):
+        matrix = np.zeros((size_X0, size_X1))
+        hash1_keys = hash1.keys()
+        hash2_keys = hash2.keys()
+        for xi in hash1_keys:
+            if xi in hash2_keys:
+                indices_1 = hash1[xi]
+                indices_2 = hash2[xi]
+                if scale == 0: 
+                    val = 1.0/len(indices_1)
+                elif scale == 1:
```

Comment:
For scale==1: I remember saying this was the right scaling but now I feel like its a bit unclear why one would want this. Maye its cleaner if we only include row-wise scaling, i.e., `1.0/len(indices_2)`. We would then have just two options for scaling (1) row-wise and (2) no-scaling.

Reply:
And then what happens with scale=0, which is used for general predict_tree_based_weights?

Reply:
Isnt it the same as 1/len(indices_1) in that general case since indices_1 == indices_2